new-log-viewer/src/contexts/UrlContextProvider.tsx

```diff
@@ -0,0 +1,215 @@
+ *
+ * @param updates An object containing key-value pairs to update the search parameters.
+ * If a key's value is `null`, the key will be removed from the search parameters.
+ * @return - The updated search string as a URLSearchParams object.
+ */
+const getUpdatedSearchParams = (updates: UrlSearchParamUpdatesType) => {
+    const newSearchParams = new URLSearchParams(window.location.search.substring(1));
+    const {filePath} = updates;
+
+    for (const [key, value] of Object.entries(updates)) {
+        if ("filePath" === key) {
+            continue;
+        }
+        if (null === value) {
+            newSearchParams.delete(key);
+        } else {
+            newSearchParams.set(key, String(value));
+        }
+    }
+    if ("string" === typeof filePath) {
+        newSearchParams.set("filePath", filePath);
+    }
+
+    return newSearchParams;
+};
+
+/**
+ * Computes updated URL search parameters based on the provided key-value pairs.
+ *
+ * @param updates An object containing key-value pairs to update the hash parameters.
+ * If a key's value is `null`, the key will be removed from the hash parameters.
+ * @return - The updated search string as a URLSearchParams object.
+ */
+const getUpdatedHashParams = (updates: UrlHashParamUpdatesType) => {
+    const newHashParams = new URLSearchParams(window.location.hash.substring(1));
+    for (const [key, value] of Object.entries(updates)) {
+        if (null === value) {
+            newHashParams.delete(key);
+        } else {
+            newHashParams.set(key, String(value));
+        }
+    }
+
+    return newHashParams;
+};
+
+/**
+ * Updates search parameters in the current window's URL based on the provided key-value pairs.
+ *
+ * @param updates An object containing key-value pairs to update the search parameters.
+ * If a key's value is `null`, the key will be removed from the search parameters.
+ */
+const updateWindowSearchParams = (updates: UrlSearchParamUpdatesType) => {
+    const newUrl = new URL(window.location.href);
+    newUrl.search = getUpdatedSearchParams(updates).toString();
+    if (!(/%23|%26/).test(newUrl.search)) {
+        newUrl.search = decodeURIComponent(newUrl.search);
+    }
+    window.history.pushState({}, "", newUrl);
+};
+
+
+/**
+ * Updates hash parameters in the current window's URL based on the provided key-value pairs.
+ *
+ * @param updates An object containing key-value pairs to update the hash parameters.
+ * If a key's value is `null`, the key will be removed from the hash parameters.
+ */
+const updateWindowHashParams = (updates: UrlHashParamUpdatesType) => {
+    const newHash = getUpdatedHashParams(updates).toString();
+    const currHash = window.location.hash.substring(1);
+    if (newHash !== currHash) {
+        const newUrl = new URL(window.location.href);
+        newUrl.hash = newHash;
+        window.history.pushState({}, "", newUrl);
+        window.dispatchEvent(new HashChangeEvent("hashchange"));
```

Comment:
This is needed because:
> pushState() never causes a hashchange event to be fired, even if the new URL differs from the old URL only in its hash.

https://developer.mozilla.org/en-US/docs/Web/API/History/pushState#description